### ISSUE TITLE
embedding_bag CPU dense backward scales gradients by the wrong frequency for mixed-count indices

### ISSUE THREAD

**Issue by bogatyy** [pytorch/pytorch#190063]
### 🐛 Describe the bug

In `_embedding_bag_dense_backward_cpu_sum_mean`, the `scale_grad_by_freq`
scaling indexes the counts array with the unique-group loop counter instead of
the index value:

https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/EmbeddingBag.cpp#L1572-L1574

```cpp
            if (scale_grad_by_freq) {
              scale /= counts[indices_data[i]];
```

`i` iterates over `next_unique_index_idx` (one step per unique index), while
the current element position is `j` and the current index value is `index`.
`indices_data[i]` therefore reads an unrelated position of the sorted indices
array. The result: a once-occurring index that sorts after a repeated one is
scaled by the repeated index's count.

```python
import torch

w = torch.ones(4, 3, requires_grad=True)
idx = torch.tensor([1, 1, 3])   # index 1 twice, index 3 once
offs = torch.tensor([0])
torch.nn.functional.embedding_bag(idx, w, offs, mode="sum",
                                  scale_grad_by_freq=True).sum().backward()
print(w.grad[:, 0].tolist())
```

Output on 2.13.0 and a source build of current main:

```text
[0.0, 1.0, 0.0, 0.5]
```

Row 1 is correct (`2 occurrences / count 2 = 1`), but row 3 should be `1.0`
(one occurrence, count 1) and is `0.5`: it was divided by `counts[indices_data[1]]
= counts[1] = 2`. The fix is `counts[index]`.

Found while implementing the corresponding MPS scaling (#190061 / #190062);
the MPS implementation divides by the true frequency, so the two backends
diverge on this input.

### Versions

PyTorch 2.14.0a0+git2662808 (source build of main) and 2.13.0; macOS 26.5.2 (arm64), Python 3.12.13. CPU-only bug, platform-independent.

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01
